com.unity.sg2/Editor/GraphDeltaRegistry/Registry/GraphType.cs

```diff
@@ -331,13 +473,20 @@ internal class GraphTypeAssignment : ICastDefinitionBuilder
 
         public bool CanConvert(FieldHandler src, FieldHandler dst)
         {
-            var srcHgt = GraphTypeHelpers.GetHeight(src);
-            var srcLen = GraphTypeHelpers.GetLength(src);
+            var srcHgt = (int)GraphTypeHelpers.GetHeight(src);
+            var srcLen = (int)GraphTypeHelpers.GetLength(src);
+
+            var dstHgt = (int)GraphTypeHelpers.GetHeight(dst);
+            var dstLen = (int)GraphTypeHelpers.GetLength(dst);
 
-            var dstHgt = GraphTypeHelpers.GetHeight(dst);
-            var dstLen = GraphTypeHelpers.GetLength(dst);
+            GraphTypeHelpers.GetDynamic(dst, out var dynLen, out var dynHgt, out _, out _);
 
-            return srcHgt == dstHgt || srcHgt == GraphType.Height.One && srcLen >= dstLen;
+            return
+                srcHgt == 1 && srcLen == 1 || // scalars can always promote to anything.
+                srcHgt == 1 && dstHgt == 1 || // vectors can truncate/zero-fill promote to other vectors.
+                // otherwise we truncate per dimension unless one of those dimensions is dynamic.
+                (srcLen >= dstLen || dynLen) && (srcHgt >= dstHgt || dynHgt);
+            // TODO: Can we just allow universal convertibility w/zero fill?
```

Comment:
Need to fix the comments here I think-- I'm not sure if `srcHgt == 1 && dstHgt == 1` case is actually necessary-- I think the dynamic rules ensure this happens where appropriate.